### PR TITLE
DEV-544: Add account settings to the AccountManager

### PR DESCRIPTION
**Pull request by Atlante45**
[JIRA ticket](https://highfidelity.atlassian.net/browse/DEV-544)

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21377/HighFidelity-Beta-PR16210-86196cf.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21377/HighFidelity-Quest-Beta-PR16210-86196cf.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21377/HighFidelity-Beta-PR16210-86196cf.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21377/HighFidelity-Beta-Interface-PR16210-86196cf.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21377/HighFidelity-Beta-PR16210-86196cf.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21377/HighFidelity-Beta-Interface-PR16210-86196cf.exe

File: libraries/networking/src/AccountManager.cpp [hunk 1]
```diff
@@ -160,33 +168,7 @@ void AccountManager::setAuthURL(const QUrl& authURL) {
 
             qCDebug(networking) << "Found metaverse API account information for" << qPrintable(_authURL.toString());
         } else {
-            // we didn't have a file - see if we can migrate old settings and store them in the new file
```
Comment (howard-stearns) [inline libraries/networking/src/AccountManager.cpp]: I'm not following what this is. Why would account-managed settings involve any changes to settings code? They should be no more connected than to, say, rendering, no?

File: libraries/networking/src/AccountManager.cpp [hunk 2]
```diff
@@ -87,6 +87,14 @@ AccountManager::AccountManager(UserAgentGetter userAgentGetter) :
...
     qRegisterMetaType<AccountManagerAuth::Type>();
     connect(this, &AccountManager::loginComplete, this, &AccountManager::uploadPublicKey);
+    connect(this, &AccountManager::loginComplete, this, &AccountManager::requestAccountSettings);
+
+    static int POST_SETTINGS_INTERVAL = 10 * MSECS_PER_SECOND;
+    _postSettingsTimer = new QTimer(this);
+    _postSettingsTimer->setInterval(POST_SETTINGS_INTERVAL);
+    connect(this, SIGNAL(loginComplete(QUrl)), _postSettingsTimer, SLOT(start()));
+    connect(this, &AccountManager::logoutComplete, _postSettingsTimer, &QTimer::stop);
+    connect(_postSettingsTimer, &QTimer::timeout, this, &AccountManager::postAccountSettings);
```
Comment (howard-stearns) [inline libraries/networking/src/AccountManager.cpp]: What happens if something gets changed on the way to shutting down? i.e., shut down sooner than 10 seconds after a change?

File: libraries/networking/src/AccountSettings.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+//    static int DEFAULT_SOME_INT;
+//    static bool DEFAULT_SOME_BOOL;
+
+    bool somethingChanged() const { return _somethingChanged; }
+
+    QJsonObject pack();
+    void unpack(QJsonObject data);
+
+//    Examples:
+//    bool hasSomeString() const { QReadLocker lock(&_settingsLock); return _hasSomeString; }
```
Comment (howard-stearns) [inline libraries/networking/src/AccountSettings.h]: I'd like the behavior to be clearer in the case that someone checks for a value or gets a value, before the data has been unpacked. It's great to have the accountSettingsLoaded signal, so  well-written app could connect to that and only read values on that signal, but....

1.  What if a script needs to get a value on startup, and the script doesn't know if you've logged in long ago, or that it got the signal during execution, or that it logged in just a moment ago but has not yet finished unpacking the data.
2. I don't want each client usage to have to figure out (or cut and paste) its own mechanism for safely getting the data. While it _should_ be up to each consumer of the data to figure out what the data means, I do think that it's up to us to give it a clean definition of what state the data is in.

File: libraries/networking/src/AccountManager.cpp (hunk 1 above)
Comment (Atlante45) [reply libraries/networking/src/AccountManager.cpp]: This is 5yo migration code from a way to store accounts long unused.
I spent a bit trying to understand it, so when I figured it out, I thought I'd clean it up.
I can pull that to a different PR if you'd prefer.

File: libraries/networking/src/AccountManager.cpp (hunk 2 above)
Comment (Atlante45) [reply libraries/networking/src/AccountManager.cpp]: It wouldn't be synced.
I should probably add a post on shutdown, but we won't have a guarantee that it goes through unless we decide to delay shutdown for it.

File: libraries/networking/src/AccountSettings.h (hunk 3 above)
Comment (Atlante45) [reply libraries/networking/src/AccountSettings.h]: Are you thinking of a something like this:
```
enum State {
    LoggedOut,
    Loading,
    Loaded
};
```

File: libraries/networking/src/AccountManager.cpp [hunk 4]
```diff
@@ -199,6 +181,10 @@ void AccountManager::setAuthURL(const QUrl& authURL) {
             refreshAccessToken();
         }
 
+        if (isLoggedIn()) {
+            emit loginComplete(_authURL);
+        }
```
Comment (Atlante45) [inline libraries/networking/src/AccountManager.cpp]: We didn't use to emit `loginComplete` when loading the account token from disk on startup, this fixes it.

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21447/HighFidelity-Beta-PR16210-1924dad.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21447/HighFidelity-Quest-Beta-PR16210-1924dad.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21447/HighFidelity-Beta-PR16210-1924dad.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21447/HighFidelity-Beta-Interface-PR16210-1924dad.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21446/HighFidelity-Beta-PR16210-b2be437.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21446/HighFidelity-Beta-Interface-PR16210-b2be437.exe

**Comment by Atlante45** [comment]
build this please

**Comment by hifi-gustavo** [comment]
Android build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21453/HighFidelity-Beta-PR16210-1924dad.apk).  Quest build is available [here](https://deployment.highfidelity.com/jobs/pr-build/label%3Dandroid/21453/HighFidelity-Quest-Beta-PR16210-1924dad.apk)

**Comment by hifi-gustavo** [comment]
Mac OS X build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21453/HighFidelity-Beta-PR16210-1924dad.dmg

**Comment by hifi-gustavo** [comment]
Mac OS X client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dmacos/21453/HighFidelity-Beta-Interface-PR16210-1924dad.dmg

**Comment by hifi-gustavo** [comment]
Windows build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21453/HighFidelity-Beta-PR16210-1924dad.exe

**Comment by hifi-gustavo** [comment]
Windows client-only build is available at: https://deployment.highfidelity.com/jobs/pr-build/label%3Dwindows/21453/HighFidelity-Beta-Interface-PR16210-1924dad.exe

**Pull request merged**